crates/biome_js_analyze/tests/specs/nursery/noSecrets/invalid.js.snap

```diff
@@ -0,0 +1,22 @@
+---
+source: crates/biome_js_analyze/tests/spec_tests.rs
+expression: invalid.js
+---
+# Input
+```jsx
+const apiKey = '12345';
+const password = 'mySecretPassword';
+var secret = "AKIA1234567890EXAMPLE"
+const slackToken = "xoxb-123456789012-123456789012-abcdefghijklmnopqrstuvwx";
+const rsaPrivateKey = "-----BEGIN RSA PRIVATE KEY-----\nMIIEpAIBAAKCAQEA1234567890..."
+const facebookToken = "facebook_app_id_12345abcde67890fghij12345";
+const twitterApiKey = "twitter_api_key_1234567890abcdefghijklmnopqrstuvwxyz";
+const githubToken = "github_pat_1234567890abcdefghijklmnopqrstuvwxyz";
+const clientSecret = "abcdefghijklmnopqrstuvwxyz"
+const herokuApiKey = "heroku_api_key_1234abcd-1234-1234-1234-1234abcd5678";
+const genericSecret = "secret_1234567890abcdefghijklmnopqrstuvwxyz";
+const genericApiKey = "api_key_1234567890abcdefghijklmnopqrstuvwxyz";
+const slackKey = "https://hooks.slack.com/services/T12345678/B12345678/abcdefghijklmnopqrstuvwx"
+const twilioApiKey = "SK1234567890abcdefghijklmnopqrstuv";
+const dbUrl = "postgres://user:password123@example.com:5432/dbname";
```

Comment:
None of these are emitting diagnostics when they should.

Reply:
Apologies. This is my first time running `just`. I think I've updated the snapshots even when they were failing, in order to try to figure out what was going wrong.

The odd part is it fails even when I run an empty `run` function:
```rs
fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        None
}
```

Even the invalid fails in this case. I don't understand what I'm doing wrong here.

Here's the stack trace:
```zsh
$ just test-lintrule noSecrets

just _touch crates/biome_js_analyze/tests/spec_tests.rs
touch crates/biome_js_analyze/tests/spec_tests.rs
just _touch crates/biome_json_analyze/tests/spec_tests.rs
touch crates/biome_json_analyze/tests/spec_tests.rs
just _touch crates/biome_css_analyze/tests/spec_tests.rs
touch crates/biome_css_analyze/tests/spec_tests.rs
just _touch crates/biome_graphql_analyze/tests/spec_tests.rs
touch crates/biome_graphql_analyze/tests/spec_tests.rs
cargo test -p biome_js_analyze -- no_secrets --show-output
   Compiling biome_js_analyze v0.5.7 (/Users/saadbazaz/Projects/biome/crates/biome_js_analyze)
warning: unused import: `regex::Regex`
 --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:5:5
  |
5 | use regex::Regex;
  |     ^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: unused variable: `ctx`
  --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:67:12
   |
67 |     fn run(ctx: &RuleContext<Self>) -> Self::Signals {
   |            ^^^ help: if this is intentional, prefix it with an underscore: `_ctx`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: constant `SENSITIVE_PATTERNS` is never used
 --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:8:7
  |
8 | const SENSITIVE_PATTERNS: &[&str] = &[
  |       ^^^^^^^^^^^^^^^^^^
  |
  = note: requested on the command line with `-W dead-code`

warning: function `is_high_entropy` is never used
  --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:83:4
   |
83 | fn is_high_entropy(text: &str) -> bool {
   |    ^^^^^^^^^^^^^^^

warning: function `calculate_shannon_entropy` is never used
  --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:89:4
   |
89 | fn calculate_shannon_entropy(data: &str) -> f64 {
   |    ^^^^^^^^^^^^^^^^^^^^^^^^^

   Compiling biome_configuration v0.0.1 (/Users/saadbazaz/Projects/biome/crates/biome_configuration)
warning: `biome_js_analyze` (lib) generated 5 warnings (run `cargo fix --lib -p biome_js_analyze` to apply 1 suggestion)
   Compiling biome_service v0.0.0 (/Users/saadbazaz/Projects/biome/crates/biome_service)
   Compiling biome_test_utils v0.0.0 (/Users/saadbazaz/Projects/biome/crates/biome_test_utils)
warning: `biome_js_analyze` (lib test) generated 5 warnings (5 duplicates)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 15.23s
     Running unittests src/lib.rs (target/debug/deps/biome_js_analyze-7f2cf2441216dc2a)

running 0 tests

successes:

successes:

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 82 filtered out; finished in 0.00s

     Running tests/quick_test.rs (target/debug/deps/quick_test-a3bdb22f6c84ea46)

running 0 tests

successes:

successes:

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 1 filtered out; finished in 0.00s

     Running tests/spec_tests.rs (target/debug/deps/spec_tests-2d02db599c4ad7ca)

running 2 tests
stored new snapshot /Users/saadbazaz/Projects/biome/crates/biome_js_analyze/tests/specs/nursery/noSecrets/valid.js.snap.new
stored new snapshot /Users/saadbazaz/Projects/biome/crates/biome_js_analyze/tests/specs/nursery/noSecrets/invalid.js.snap.new
test specs::nursery::no_secrets::valid_js ... FAILED
test specs::nursery::no_secrets::invalid_js ... FAILED

successes:

successes:

failures:

---- specs::nursery::no_secrets::valid_js stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: crates/biome_js_analyze/tests/specs/nursery/noSecrets/valid.js.snap
Snapshot: valid.js
Source: crates/biome_js_analyze/tests/spec_tests.rs:86
──────────────────────────────────────────────────────────────────────────────────────────────────
Expression: valid.js
──────────────────────────────────────────────────────────────────────────────────────────────────
+new results
────────────┬─────────────────────────────────────────────────────────────────────────────────────
          0 │+# Input
          1 │+```jsx
          2 │+var a = 1;
          3 │+const userName = 'testUser';
          4 │+let count = 10;
          5 │+var nonSecret = "hello world"
          6 │+```
────────────┴─────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'specs::nursery::no_secrets::valid_js' panicked at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/runtime.rs:562:9:
snapshot assertion for 'valid.js' failed in line 86
stack backtrace:
   0: rust_begin_unwind
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/panicking.rs:72:14
   2: insta::runtime::finalize_assertion
             at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/runtime.rs:562:9
   3: insta::runtime::assert_snapshot
             at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/runtime.rs:690:9
   4: spec_tests::run_test::{{closure}}
             at ./tests/spec_tests.rs:86:9
   5: insta::settings::Settings::bind
             at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/settings.rs:510:9
   6: spec_tests::run_test
             at ./tests/spec_tests.rs:82:5
   7: spec_tests::specs::nursery::no_secrets::valid_js
             at ./tests/spec_tests.rs:16:1
   8: spec_tests::specs::nursery::no_secrets::valid_js::{{closure}}
             at ./tests/spec_tests.rs:16:111
   9: core::ops::function::FnOnce::call_once
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/ops/function.rs:250:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- specs::nursery::no_secrets::invalid_js stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: crates/biome_js_analyze/tests/specs/nursery/noSecrets/invalid.js.snap
Snapshot: invalid.js
Source: crates/biome_js_analyze/tests/spec_tests.rs:86
──────────────────────────────────────────────────────────────────────────────────────────────────
Expression: invalid.js
──────────────────────────────────────────────────────────────────────────────────────────────────
+new results
────────────┬─────────────────────────────────────────────────────────────────────────────────────
          0 │+# Input
          1 │+```jsx
          2 │+const apiKey = '12345';
          3 │+const password = 'mySecretPassword';
          4 │+var secret = "AKIA1234567890EXAMPLE"
          5 │+const slackToken = "xoxb-123456789012-123456789012-abcdefghijklmnopqrstuvwx";
          6 │+const rsaPrivateKey = "-----BEGIN RSA PRIVATE KEY-----\nMIIEpAIBAAKCAQEA1234567890..."
          7 │+const facebookToken = "facebook_app_id_12345abcde67890fghij12345";
          8 │+const twitterApiKey = "twitter_api_key_1234567890abcdefghijklmnopqrstuvwxyz";
          9 │+const githubToken = "github_pat_1234567890abcdefghijklmnopqrstuvwxyz";
         10 │+const clientSecret = "abcdefghijklmnopqrstuvwxyz"
         11 │+const herokuApiKey = "heroku_api_key_1234abcd-1234-1234-1234-1234abcd5678";
         12 │+const genericSecret = "secret_1234567890abcdefghijklmnopqrstuvwxyz";
         13 │+const genericApiKey = "api_key_1234567890abcdefghijklmnopqrstuvwxyz";
         14 │+const slackKey = "https://hooks.slack.com/services/T12345678/B12345678/abcdefghijklmnopqrstuvwx"
         15 │+const twilioApiKey = "SK1234567890abcdefghijklmnopqrstuv";
         16 │+const dbUrl = "postgres://user:password123@example.com:5432/dbname";
         17 │+```
────────────┴─────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'specs::nursery::no_secrets::invalid_js' panicked at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/runtime.rs:562:9:
snapshot assertion for 'invalid.js' failed in line 86
stack backtrace:
   0: rust_begin_unwind
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/std/src/panicking.rs:652:5
   1: core::panicking::panic_fmt
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/panicking.rs:72:14
   2: insta::runtime::finalize_assertion
             at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/runtime.rs:562:9
   3: insta::runtime::assert_snapshot
             at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/runtime.rs:690:9
   4: spec_tests::run_test::{{closure}}
             at ./tests/spec_tests.rs:86:9
   5: insta::settings::Settings::bind
             at /Users/saadbazaz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.39.0/src/settings.rs:510:9
   6: spec_tests::run_test
             at ./tests/spec_tests.rs:82:5
   7: spec_tests::specs::nursery::no_secrets::invalid_js
             at ./tests/spec_tests.rs:16:1
   8: spec_tests::specs::nursery::no_secrets::invalid_js::{{closure}}
             at ./tests/spec_tests.rs:16:111
   9: core::ops::function::FnOnce::call_once
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/ops/function.rs:250:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/051478957371ee0084a7c0913941d2a8c4757bb9/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    specs::nursery::no_secrets::invalid_js
    specs::nursery::no_secrets::valid_js

test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 1051 filtered out; finished in 0.30s

error: test failed, to rerun pass `-p biome_js_analyze --test spec_tests`
error: Recipe `test-lintrule` failed on line 128 with exit code 101
```

Reply:
These are snapshot test failures. Use `cargo insta accept` to accept the changes to the snapshot.

But you'll still need to make sure all the lines in `invalid.js` get flagged by this rule. There should be a section in the snapshot that starts with `# Diagnostics`.

Reply:
Done! Thanks so much, that helped me get it actually working too :p Seems like I was doing something wrong.

Reply:
Check https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#snapshots for more information.
Besides that do not forget to run `just f` `just l` and `just gen-lint`

Reply:
Done @minht11 .
While I work on the refactoring, could you and @dyc3 confirm if the tests look alright to you?